### PR TITLE
Outline multiple ops in HLO to CUDNN lowering

### PR DESCRIPTION
**Pull request by chsigg**

### CODE REVIEW

**Review by frgossen — COMMENTED** [review]
Thanks!

File: compiler/src/openxla/compiler/nvgpu/Conversion/ConvertHLOToCUDNN.cpp [hunk 1]
```diff
@@ -52,15 +59,19 @@ static FailureOr<Layout> getCudnnKernelLayout(int64_t inputDim,
                                               int64_t outputDim,
                                               MLIRContext* ctx) {
   if (outputDim != 0) return failure();
-  if (inputDim == 1) return Layout::KCHW;
-  if (inputDim == 3) return Layout::KHWC;
+  // Return input/output layout instead of the kernel layout so that casts
+  // cancel themselves. The actual layouts are the same and the lowering and
+  // runtime implementation handle it correctly. Eventually, this should be
+  // replaced by layout propagation.
```
[the remark this answers is not in the document]
Comment (chsigg) [reply compiler/src/openxla/compiler/nvgpu/Conversion/ConvertHLOToCUDNN.cpp]: Done.

**Review by ezhulenev — APPROVED** [review]

File: compiler/src/openxla/compiler/nvgpu/Conversion/ConvertHLOToCUDNN.cpp [hunk 2]
```diff
@@ -201,34 +212,56 @@ static FailureOr<ConvLayouts> matchConv(stablehlo::ConvolutionOp op,
...
+    }
+    ++i;
+  }
+  if (arguments.size() == 1 && arguments.front() == result) return failure();
+  auto getType = [&](TensorValue operand) -> Type {
+    return getCudnnTensorType(operand.getType(), layout.value_or(Layout::NCHW));
+  };
+  FunctionType funcType = rewriter.getFunctionType(
+      to_vector(map_range(arguments, getType)), getType(result));
+  return outlineToGraph(result, arguments, funcType, rewriter);
```
Comment (ezhulenev) [inline compiler/src/openxla/compiler/nvgpu/Conversion/ConvertHLOToCUDNN.cpp:255]: At some point outlining has to check that all intermediates has one use (or better no uses outside of an outlined graph), I don't think cuDNN supports materializing intermediate results into buffers.

Comment (chsigg) [reply compiler/src/openxla/compiler/nvgpu/Conversion/ConvertHLOToCUDNN.cpp:255]: Good point, I forgot about that. I will fix this in a follow-up.

**Pull request merged**